### PR TITLE
Entrust plugin

### PR DESCRIPTION
**Pull request by sirferl**
I created the plugin for the Entrust API. I added it to the setup.py. Configuration documentation is also provided. Please merge this.

### CODE REVIEW

**Review by hosseinsh — COMMENTED** [review]
@sirferl thanks for the PR. 
I have left some feedback on how to improve this plugin, and happy to work with you to get it there.

it would be really great, if we could also add some testing, like the digiCert plugin
https://github.com/Netflix/lemur/blob/e4ed2a14e7aa4621187279e4015e916081514f48/lemur/plugins/lemur_digicert/tests/test_digicert.py

**Comment by sirferl** [comment]
Hi @hosseinsh, I just wanted to confirm that I will work on the suggestions. ferdinand

File: lemur/plugins/lemur_entrust/plugin.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    title = "ENTRUST"
+    slug = "entrust-issuer"
+    description = "Enables the creation of certificates by ENTRUST"
+    version = ENTRUST.VERSION
+
+    author = "sirferl"
+    author_url = "https://github.com/sirferl/lemur"
+
+    def __init__(self, *args, **kwargs):
+        """Initialize the issuer with the appropriate details."""
```
[the remark this answers is not in the document]
Comment (sirferl) [reply lemur/plugins/lemur_entrust/plugin.py:64]: Added this check

File: lemur/plugins/lemur_entrust/plugin.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        :return:
+        """
+        entrust_root = current_app.config.get("ENTRUST_ROOT")
+        entrust_issuing = current_app.config.get("ENTRUST_ISSUING")
+        role = {"username": "", "password": "", "name": "entrust"}
+        current_app.logger.info("Creating Auth: {0} {1}".format(options, entrust_issuing))
+        return entrust_root, "", [role]
+
+    def revoke_certificate(self, certificate, comments):
+        raise NotImplementedError("Not implemented\n", self, certificate, comments)
```
[the remark this answers is not in the document]
Comment (sirferl) [reply lemur/plugins/lemur_entrust/plugin.py]: I added the revocation functionality and a response handler like suggested!

File: lemur/plugins/lemur_entrust/plugin.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    author = "sirferl"
+    author_url = "https://github.com/sirferl/lemur"
+
+    def __init__(self, *args, **kwargs):
+        """Initialize the issuer with the appropriate details."""
+        self.session = requests.Session()
+        cert_file_path = current_app.config.get("ENTRUST_API_CERT")
+        key_file_path = current_app.config.get("ENTRUST_API_KEY")
+        user = current_app.config.get("ENTRUST_API_USER")
+        passw = current_app.config.get("ENTRUST_API_PASS")
```
[the remark this answers is not in the document]
Comment (sirferl) [reply lemur/plugins/lemur_entrust/plugin.py:108]: followed your suggestion here.

File: lemur/plugins/lemur_entrust/plugin.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        """
+        current_app.logger.info(
+            "Requesting options: {0}".format(issuer_options)
+        )
+
+        url = current_app.config.get("ENTRUST_URL") + "/certificates"
+
+        data = process_options(issuer_options)
+        data["csr"] = csr
+        current_req = arrow.utcnow().format('YYYY-MM-DD HH:mm:ss')
```
[the remark this answers is not in the document]
Comment (sirferl) [reply lemur/plugins/lemur_entrust/plugin.py]: This was a residual from timeout error hunting. I had to mark the requests to see which one returns data and which one times out. I removed it.

File: lemur/plugins/lemur_entrust/plugin.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    expiry_date = arrow.utcnow().shift(years=1, days=+10).format('YYYY-MM-DD')
+
+    tracking_data = {
+        "requesterName": current_app.config.get("ENTRUST_NAME"),
+        "requesterEmail": current_app.config.get("ENTRUST_EMAIL"),
+        "requesterPhone": current_app.config.get("ENTRUST_PHONE")
+    }
+
+    data = {
+        "signingAlg": "SHA-2",
```
[the remark this answers is not in the document]
Comment (sirferl) [reply lemur/plugins/lemur_entrust/plugin.py:45]: Hi, entrust only supports "SHA-2" at the moment. I'll leave it at that. If this should change in the future I can change the code.

File: lemur/plugins/lemur_entrust/plugin.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+    :param options:
+    :return: dict of valid entrust options
+    """
+    # if there is a config variable ENTRUST_PRODUCT_<upper(authority.name)>
+    # take the value as Cert product-type
+    # else default to "STANDARD_SSL"
+    authority = options.get("authority").name.upper()
+    product_type = current_app.config.get("ENTRUST_PRODUCT_{0}".format(authority), "STANDARD_SSL")
+    expiry_date = arrow.utcnow().shift(years=1, days=+10).format('YYYY-MM-DD')
```
[the remark this answers is not in the document]
Comment (sirferl) [reply lemur/plugins/lemur_entrust/plugin.py:36]: Entrust just allows a year + one month. A starting date is not allowed. I added code similarly to your suggestion that takes this into account. I tried to enter  an end date in the UI I encountered an error. Maybe we should raise an issue about that. Error message was: "Schema - {"rule":["If validity start is specified so must validity end."]}"

File: lemur/plugins/lemur_entrust/plugin.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    title = "ENTRUST"
+    slug = "entrust-source"
+    description = "Enables the collecion of certificates"
+    version = ENTRUST.VERSION
+
+    author = "sirferl"
+    author_url = "https://github.com/sirferl/lemur"
+    options = [
+        {
+            "name": "dummy",
```
[the remark this answers is not in the document]
Comment (sirferl) [reply lemur/plugins/lemur_entrust/plugin.py:202]: I removed that. I remember I had an error once when no options where specified.

**Comment by hosseinsh** [comment]
Thanks @sirferl.

File: lemur/plugins/lemur_entrust/plugin.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hosseinsh) [reply lemur/plugins/lemur_entrust/plugin.py:66]: yeah, I noticed that in their docs.

File: lemur/plugins/lemur_entrust/plugin.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (hosseinsh) [reply lemur/plugins/lemur_entrust/plugin.py]: a starting date would be a feature request to raise with Entrust.

**Comment by sirferl** [comment]
@hosseinsh could you merge it in. I want to set up a production environment with the new plugin...I will add the test as well.

**Comment by hosseinsh** [comment]
Thanks @sirferl for addressing the comments, looking forward to the tests and good luck with deploying the plugin to prod.

**Comment by sirferl** [comment]
thank you. I will deliver the test-script, but I've seen there are some things I have to learn first...

**Comment by hosseinsh** [comment]
cool, no worries, I can also help with some parts.

**Pull request merged**